### PR TITLE
Introduce serialization for actions

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request

This PR builds on the `ActionMap` PR (#6900) by leveraging `ActionMap` to serialize actions. From the top down, the process is now as follows:
- `CascadiaSettings`: remove the hack of copying whatever we had for actions before.
- `GlobalAppSettings`: serialize the `ActionMap` to `"actions": []`
- `ActionMap`: iterate over the internal `_ActionMap` (list of actions) and serialize each `Command`
- `Command`: **THIS IS WHERE THE MAGIC HAPPENS!** For _each_ key mapping, serialize an action. Only the first one needs to include the name and icon.
- `ActionAndArgs`: Find the relevant `IActionArgs` parser and serialize the `ActionAndArgs`.
- `ActionArgs`: **ANNOYING CHANGE** Serialize any args that are set. We _need_ each setting to be saved as a `std::optional`. As with inheritance, this allows us to distinguish an explicit setting to the default value (sometimes `null`) vs an implicit "give me the default value". This allows us to serialize only the relevant details of each action, rather than writing _all_ of the args.

## References
- #8100: Inheritance/Layering for lists
   - This tracks layering and better serialization for color schemes _and_ actions. This PR resolves half of that issue. The next step is to apply the concepts used in this PR (and #9621) to solve the similar problem for color schemes.
- #6900: Actions page

## Validation Steps Performed
Tests added!

### CODE REVIEW

**Comment by DHowett** [comment]
I have this ...

```json
        {
            "commands": 
            [
                {
                    "command": 
                    {
                        "action": "sendInput",
                        "input": "${profile.name}"
                    },
                    "iterateOn": "profiles"
                }
            ],
            "name": "Send Input ..."
        },
        {
            "commands": 
            [
                {
                    "commands": 
                    [
                        {
                            "command": 
                            {
                                "action": "sendInput",
                                "input": "${profile.name} ${scheme.name}"
                            },
                            "iterateOn": "schemes"
                        }
                    ],
                    "iterateOn": "profiles",
                    "name": "nest level (${profile.name})"
                }
            ],
            "name": "Send Input (Evil) ..."
        }
```

and this pull request rewrote it (!) as this:

```json
        {
            "command": "unbound",
            "name": "Send Input (Evil) ..."
        },
        {
            "command": "unbound",
            "name": "Send Input ..."
        }
```

**Comment by carlos-zamora** [comment]
> I have this ...

Welp... It was a one liner hahaha.

File: src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp [hunk 1]
```diff
@@ -49,6 +56,30 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
...
     }
 
+    Json::Value ActionMap::ToJson() const
+    {
+        Json::Value actionList{ Json::ValueType::arrayValue };
+        for (const auto& [_, cmd] : _ActionMap)
+        {
+            const auto cmdImpl{ winrt::get_self<implementation::Command>(cmd) };
+            const auto& cmdJsonArray{ cmdImpl->ToJson() };
+            for (const auto& cmdJson : cmdJsonArray)
```
Comment (zadjii-msft) [inline src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp:66]: So wait, if I had 
```jsonc  
{
    "icon": "whatever icon",
    "name": "whatever name...",
    "commands": [
        {
            "command": { "action": "splitPane", "profile": "a name 1", "split": "auto" }
        },
        {
            "command": { "action": "splitPane", "profile": "a name 1", "split": "vertical" }
        },
        {
            "command": { "action": "splitPane", "profile": "a name 1", "split": "horizontal" }
        }
    ]
}
```

(note that this is _not_ an iterable command)

originally, is that going to:
* first get serialized as-is, by the `if (_nestedCommand)` branch in `Command::ToJson`, THEN
* get serialized _again_ by the `for (const auto& [_, cmd] : _NestedCommands)` loop below as 
```json
{
    "command": { "action": "splitPane", "profile": "a name 1", "split": "auto" }
},
{
    "command": { "action": "splitPane", "profile": "a name 2", "split": "vertical" }
},
{
    "command": { "action": "splitPane", "profile": "a name 3", "split": "horizontal" }
}
```

For a total of 4 top-level commands: one with three nested beneath it, and 3 more top-level commands?

Or is the original top-level command (`whatever name...`) only in `_NestedCommands`, so it's not an issue?

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
I only really have the one main question that I want answered. Otherwise, the rest seems pretty straightforward.

We should make sure to add a test for https://github.com/microsoft/terminal/pull/9926#issuecomment-832918162 as well, because it's a good edge case

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp [hunk 2]
```diff
@@ -166,6 +129,34 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         { ShortcutAction::Invalid, nullptr },
     };
 
+    using ActionArgSerializerFunction = std::function<Json::Value(IActionArgs)>;
+    static const std::map<ShortcutAction, ActionArgSerializerFunction, std::less<>> argSerializers{
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:137]: I _porque no los dos_-ed it.

One day, we'll replace all of the `FromJson` and `ToJson` to be inside `JsonUtils::ConversionTrait`s. So, I'm hoping that'll make all of this look a bit nicer when that time comes.

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp [hunk 3]
```diff
@@ -271,6 +261,49 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
...
+                // "command": "copy"
+                const auto shortcutActionIter{ ActionToStringMap.find(val.Action()) };
+                if (shortcutActionIter != ActionToStringMap.end())
+                {
+                    return JsonKey(shortcutActionIter->second);
+                }
+            }
+            else
+            {
+                // "command": { "action": "copy", "singleLine": false }
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:279]: 1. `"command": { "action": "copy" }`

https://github.com/microsoft/terminal/blob/b506817508da9a4d5cd403a9ce1ba9f738eb65d6/src/cascadia/LocalTests_SettingsModel/SerializationTests.cpp#L240-L243

We _could_ make it return `2` by...
1. check if all of the members of the `IActionArgs` are `std::nullopt`.
2. If true, write the shorthand version

But step `1` of that would be really annoying to implement. It could be a bit less annoying to implement given `JsonConverters`, but I digress.

File: src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp (hunk 1 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp:66]: Nah, we only do it once. Here's what the trace looks like:
1. `ActionMap::AddAction()`
   - if the `Command` is the root of a nested command, add the root to `_NestedCommands`. We DO NOT add the sub commands to `_ActionMap`.
   - otherwise, add it to `_ActionMap`
2. `ActionMap::ToJson()`
   - serialize all of the `_ActionMap` commands
      3. `Command::ToJson` --> <more complicated, out of scope of this question>
   - serialize all of the `_NestedCommands` commands
      3. `Command::ToJson` --> just spit out the `Command::_originalJson`

So since we don't actually record the sub-commands in `ActionMap`, we don't ever write them out other than via `Command::_originalJson`.

NOTE: _theoretically_, we could just always return `Command::_originalJson`. But I'm sensing that we shouldn't depend on that.

**Review by zadjii-msft — APPROVED** [review]
🙌 on the `ACTION_TO_SERIALIZERS_PAIR` thing, and thanks for adding the test!

**Comment by carlos-zamora** [comment]
> Move focus doesn't display direction

This bug was caused by the list of `ALL_SHORTCUT_ACTIONS_WITH_ARGS` missing `MoveFocus`. I went ahead and sorted it, then double checked that I'm not missing anything.

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp [hunk 4]
```diff
@@ -78,98 +83,38 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
...
-        { OpenWindowRenamerKey, ShortcutAction::OpenWindowRenamer },
-        { GlobalSummonKey, ShortcutAction::GlobalSummon },
-        { QuakeModeKey, ShortcutAction::QuakeMode },
+
+#define ON_ALL_ACTIONS(action) KEY_TO_ACTION_PAIR(action)
+        ALL_SHORTCUT_ACTIONS
+#undef ON_ALL_ACTIONS
+    };
+
+    static const std::map<ShortcutAction, std::string_view, std::less<>> ActionToStringMap{
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:92]: We have got to stop it with the globally initialized maps. Sorry. This looks like a case for til::static_map or til::presorted_static_map. I'd rather do that than have yet another static initializer :smile:

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp [hunk 5]
```diff
@@ -271,6 +216,49 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
...
+    {
+        if (val)
+        {
+            if (!val.Args())
+            {
+                // "command": "copy"
+                const auto shortcutActionIter{ ActionToStringMap.find(val.Action()) };
+                if (shortcutActionIter != ActionToStringMap.end())
+                {
+                    return JsonKey(shortcutActionIter->second);
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:229]: please avoid Jsonkey if possible :|

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp [hunk 6]
```diff
@@ -271,6 +216,49 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
...
+                    if (pfn)
+                    {
+                        result = pfn(val.Args());
+                    }
+                }
+
+                // Set the "action" part
+                const auto shortcutActionIter{ ActionToStringMap.find(val.Action()) };
+                result[JsonKey(ActionKey)] = shortcutActionIter == ActionToStringMap.end() ?
+                                                 JsonKey(UnboundKey) :
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:251]: this seems weird actually. The code flow suggests that it might accidentally produce a `{action: unbound, copySingleLine: true}` because... we don't _skip it entirely_ when it isn't found in `ActionToStringMap`

File: src/cascadia/TerminalSettingsModel/ActionArgs.cpp [hunk 7]
```diff
@@ -257,7 +257,7 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
     winrt::hstring MoveFocusArgs::GenerateName() const
     {
         winrt::hstring directionString;
-        switch (_FocusDirection)
+        switch (FocusDirection())
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionArgs.cpp:260]: why did all these have to change?

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/ActionArgs.cpp:260]: oh, because of the change to optionals i guess?

File: src/cascadia/TerminalSettingsModel/ActionArgs.h [hunk 8]
```diff
@@ -489,12 +559,21 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
...
+            if (args->SplitSize() >= 1 || args->SplitSize() <= 0)
             {
                 return { nullptr, { SettingsLoadWarnings::InvalidSplitSize } };
             }
             return { *args, {} };
         }
+        static Json::Value ToJson(const IActionArgs& val)
+        {
+            const auto args{ get_self<SplitPaneArgs>(val) };
+            auto json{ NewTerminalArgs::ToJson(args->_TerminalArgs) };
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionArgs.h:571]: what if it's nullptr? be careful!

File: src/cascadia/TerminalSettingsModel/ActionArgs.h [hunk 9]
```diff
@@ -474,7 +544,7 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
             if (otherAsUs)
             {
                 return otherAsUs->_SplitStyle == _SplitStyle &&
-                       (otherAsUs->_TerminalArgs ? otherAsUs->_TerminalArgs.Equals(_TerminalArgs) :
+                       (otherAsUs->_TerminalArgs ? otherAsUs->_TerminalArgs.Equals(TerminalArgs()) :
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionArgs.h:547]: Why does this Equals have to use the fully realized value for TerminalArgs() when it isn't even an optional? Be careful here!

File: src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp [hunk 10]
```diff
@@ -49,6 +56,30 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         return warnings;
     }
 
+    Json::Value ActionMap::ToJson() const
+    {
+        Json::Value actionList{ Json::ValueType::arrayValue };
+        for (const auto& [_, cmd] : _ActionMap)
+        {
+            const auto cmdImpl{ winrt::get_self<implementation::Command>(cmd) };
+            const auto& cmdJsonArray{ cmdImpl->ToJson() };
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp:65]: what? a command serializes to multiple instances?? why? How do we validate that contract? It's just called `ToJson` :|

File: src/cascadia/TerminalSettingsModel/Command.cpp [hunk 11]
```diff
@@ -383,6 +384,65 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
...
+            {
+                cmdJson[JsonKey(ActionKey)] = ActionAndArgs::ToJson(_ActionAndArgs);
+            }
+
+            cmdList.append(cmdJson);
+        }
+        else
+        {
+            // we'll write out one command per key mapping
+            for (auto keys{ _keyMappings.begin() }; keys != _keyMappings.end(); ++keys)
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/Command.cpp:422]:
```suggestion
            for (const auto keys{ _keyMappings.begin() }; keys != _keyMappings.end(); ++keys)
```

**Review by DHowett — CHANGES_REQUESTED** [review]

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp (hunk 4 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:92]: Should I do what we did with `GeneratedActionNames` in `ActionAndArgs::GenerateName`? Define these in the function they're used as static lambda functions. Or does that not make this any better?

**Comment by DHowett** [comment]
The problem is the static map, not the functions.

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp (hunk 4 above)
Comment (lhecker) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:92]: @DHowett Can you explain the problem with static initializers for the uninitiated? 🙂

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:92]: Sure!

Initializing a static _non-constexpr_ value at file scope generates a block of code that must be executed when the library is loaded, every time it is loaded. For something like the settings model, that cost is paid once on startup, but it is still a cost that _must_ be paid before any of the actual settings model code is run.

constexpr is a different story: those are evaluated at compile time and stored permanently in the binary.

For a really, really good example of the cost of a static initializer (or something very similar) and the payback you get from replacing it with a fully-constexpr data store, look at commit 16e1e29a12ac3f3f2f41d0a24be41dedab1f5605.

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp (hunk 6 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:251]: Refactored it a bit to make it more clear.

File: src/cascadia/TerminalSettingsModel/ActionArgs.cpp (hunk 7 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionArgs.cpp:260]: Yup! It was annoying, but definitely for the better.

File: src/cascadia/TerminalSettingsModel/ActionArgs.h (hunk 9 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionArgs.h:547]: oh huh, that was the only place I did that. Good catch!

File: src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp (hunk 10 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp:65]: Added the following. Hope it clears it up. :)
```c++
// Command serializes to an array of JSON objects.
            // This is because a Command may have multiple key chords associated with it.
            // The name and icon are only serialized in the first object.
            // Example:
            // { "name": "Custom Copy", "command": "copy", "keys": "ctrl+c" }
            // {                        "command": "copy", "keys": "ctrl+shift+c" }
            // {                        "command": "copy", "keys": "ctrl+ins" }
```

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionMapSerialization.cpp:65]: There's a serialization test that covers this case:
```js
{
	[
		{ "command": { "action": "adjustFontSize", "delta": 1 }, "keys": "ctrl+c" },
		{ "command": { "action": "adjustFontSize", "delta": 1 }, "keys": "ctrl+d" }
	]
}
```

File: src/cascadia/TerminalSettingsModel/Command.cpp (hunk 11 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/Command.cpp:422]: I'll do you one better:
```cpp
for (const auto& keys : _keyMappings)
```

File: src/cascadia/TerminalSettingsModel/Command.cpp [hunk 12]
```diff
@@ -419,11 +419,11 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
         else
         {
             // we'll write out one command per key mapping
-            for (auto keys{ _keyMappings.begin() }; keys != _keyMappings.end(); ++keys)
+            for (const auto& keys : _keyMappings)
             {
                 Json::Value cmdJson{ Json::ValueType::objectValue };
 
-                if (keys == _keyMappings.begin())
+                if (keys == *_keyMappings.begin())
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/Command.cpp:426]: absolutely do not do this ;P

what you had before was correct -- you want to check "am I at the beginning of the list?". You do not want to check, "am I at an item that is equivalent to the item at the beginning of the list" (!!)

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp (hunk 4 above)
Comment (lhecker) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:92]: This seems fairly solid - it even offers perfect hash functions: https://github.com/serge-sans-paille/frozen

File: src/cascadia/TerminalSettingsModel/Command.cpp (hunk 11 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/Command.cpp]: BTW, `const` didn't work on this because of the `++keys` at the end

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp (hunk 4 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:92]: Leonard and I couldn't get this to work, unfortunately. We kept running into `til::presorted_static_map` saying that it couldn't find the `static_map` in the ctor or something like that. Sorry!

**Review by DHowett — APPROVED** [review]
